### PR TITLE
C++ API parity: Dropout, Dropout2d, Dropout3d

### PR DESCRIPTION
**Pull request by pbelevich**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29761 C++ API parity: Dropout, Dropout2d, Dropout3d**

Differential Revision: [D18530820](https://our.internmc.facebook.com/intern/diff/D18530820)

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/options/dropout.h [hunk 1]
```diff
@@ -7,14 +7,41 @@
...
+  TORCH_ARG(bool, inplace) = false;
+
+  TORCH_ARG(bool, training) = false;
 };
 
+using Dropout2dFuncOptions = DropoutFuncOptions;
+
+using Dropout3dFuncOptions = DropoutFuncOptions;
+
+}
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/dropout.h]: nit:
```suggestion
} // namespace functional
```

File: torch/csrc/api/include/torch/nn/options/dropout.h [hunk 2]
```diff
@@ -7,14 +7,41 @@
...
+
+/// Options for `Dropout` functional.
+struct TORCH_API DropoutFuncOptions {
+  /// The probability of an element to be zeroed. Default: 0.5
+  TORCH_ARG(double, p) = 0.5;
+
+  /// can optionally do the operation in-place. Default: False
+  TORCH_ARG(bool, inplace) = false;
+
+  TORCH_ARG(bool, training) = false;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/dropout.h]: Based on https://github.com/pytorch/pytorch/blob/3fb9bbc99b8e1af286ca84045697a951f5346832/torch/nn/functional.py#L788, it seems that the default value for `training` should be `true`

Comment (yf225) [reply torch/csrc/api/include/torch/nn/options/dropout.h]: also it would be awesome to move `TORCH_ARG(bool, training) = false;` above `TORCH_ARG(bool, inplace) = false;`, to be consistent with the argument order of the Python functional `dropout` :D

File: torch/csrc/api/include/torch/nn/modules/dropout.h [hunk 3]
```diff
@@ -14,66 +14,90 @@ namespace torch {
...
  public:
-  explicit DropoutImplBase(const DropoutOptions& options_ = DropoutOptions());
+  _DropoutNd(double p) : _DropoutNd(DropoutOptions().p(p)) {};
+
+  explicit _DropoutNd(const DropoutOptions& options_ = {})
+    : options(options_) {
+    TORCH_CHECK(
+        options.p() >= 0. && options.p() <= 1.,
+        "dropout probability has to be between 0 and 1, but got ",
+        options.p());
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/dropout.h]: We can probably move the check to the `reset()` function and call `reset()` from the constructor, which would be more consistent with the design of the other C++ modules :D

File: torch/csrc/api/include/torch/nn/functional/dropout.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  if (inplace) {
+    return torch::dropout_(input, p, training);
+  } else {
+    return torch::dropout(input, p, training);
+  }
+}
+
+} // namespace detail
+
+inline Tensor dropout(Tensor &input,
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/dropout.h]: nit:
```suggestion
inline Tensor dropout(Tensor& input, 
```

File: torch/csrc/api/include/torch/nn/functional/dropout.h [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  if (inplace) {
+    return torch::feature_dropout_(input, p, training);
+  } else {
+    return torch::feature_dropout(input, p, training);
+  }
+}
+
+} // namespace detail
+
+inline Tensor dropout2d(Tensor &input,
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/dropout.h]: nit:
```suggestion
inline Tensor dropout2d(Tensor& input, 
```

File: torch/csrc/api/include/torch/nn/functional/dropout.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  if (inplace) {
+    return torch::feature_dropout_(input, p, training);
+  } else {
+    return torch::feature_dropout(input, p, training);
+  }
+}
+
+} // namespace detail
+
+inline Tensor dropout3d(Tensor &input,
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/dropout.h]: nit:
```suggestion
inline Tensor dropout3d(Tensor& input, 
```

File: torch/csrc/api/include/torch/nn/functional/dropout.h [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+inline Tensor dropout3d(Tensor &input, 
+    const Dropout3dFuncOptions& options = {}) {
+  return detail::dropout3d(
+      input, options.p(), options.training(), options.inplace());
+}
+
+} // namespace functional
+} // namespace nn
+} // namespace torch
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/dropout.h]: nit: we might need a new line here

File: torch/csrc/api/include/torch/nn/modules/dropout.h [hunk 8]
```diff
@@ -14,66 +14,90 @@ namespace torch {
...
+/// See the documentation for `Dropout3dImpl` class to learn what methods it
+/// provides, or the documentation for `ModuleHolder` to learn about PyTorch's
+/// module storage semantics.
+TORCH_MODULE(Dropout3d);
 
-/// A `ModuleHolder` subclass for `FeatureDropoutImpl`.
-/// See the documentation for `FeatureDropoutImpl` class to learn what methods
-/// it provides, or the documentation for `ModuleHolder` to learn about
-/// PyTorch's module storage semantics.
-TORCH_MODULE(FeatureDropout);
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/dropout.h]: I think we might want to deprecate `FeatureDropout` for one release and then remove it in the next one. For 1.4 (which will be branch cut on 11/20) we can deprecate it first by printing a warning message in the `FeatureDropout` constructor asking people to move to `Dropout2d` / `Dropout3d`, and in the 1.5 release we can remove `FeatureDropout` completely :D

**Review by yf225 — COMMENTED** [review]
@pbelevich Thanks a lot for the awesome work! I left some comments

File: torch/csrc/api/src/nn/modules/dropout.cpp [hunk 9]
```diff
@@ -8,42 +9,71 @@
...
+  : detail::_DropoutNd<FeatureDropoutImpl>(p) {
+  TORCH_WARN("torch::nn::FeatureDropout module is deprecated."
+             "Use Dropout{2,3}d instead.");
+}
+
+FeatureDropoutImpl::FeatureDropoutImpl(const FeatureDropoutOptions& options_)
+  : detail::_DropoutNd<FeatureDropoutImpl>(options_) {
+  TORCH_WARN("torch::nn::FeatureDropout module is deprecated."
+             "Use Dropout{2,3}d instead.");
+}
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/dropout.cpp]: It would be awesome to test for the deprecation message, similar to https://github.com/pytorch/pytorch/blob/eef349a679f5331e05de3c9462e285d1ab99c8a8/test/cpp/api/modules.cpp#L1176-L1188

File: torch/csrc/api/include/torch/nn/modules/dropout.h [hunk 10]
```diff
@@ -14,34 +14,95 @@ namespace torch {
...
+
+  /// Pretty prints the `Dropout3d` module into the given `stream`.
+  void pretty_print(std::ostream& stream) const override;
+};
+
+/// A `ModuleHolder` subclass for `Dropout3dImpl`.
+/// See the documentation for `Dropout3dImpl` class to learn what methods it
+/// provides, or the documentation for `ModuleHolder` to learn about PyTorch's
+/// module storage semantics.
+TORCH_MODULE(Dropout3d);
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/dropout.h]: I believe we can update the parity status of `Dropout` / `Dropout2d` / `Dropout3d` in `test/cpp_api_parity/parity-tracker.md` as well :D

**Review by yf225 — APPROVED** [review]
@pbelevich Thanks so much for the awesome work!
